vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+        size_t max_nodes;
+        size_t leaf_example_multiplier;
+        size_t max_routers;
+        size_t max_num_labels;
+        float alpha; //for cpt type of update.
+        size_t routers_used;
+        int iter;
+        uint32_t dream_repeats; //number of dream operations per example.
+
+        uint32_t total_num_queires; 
+
+        size_t max_depth;
+        size_t max_ex_in_leaf;
+
+        float construct_time;
+        float test_time;
+        float cumulative_reward;
+        float cumulative_reward_count;
+
+        bool path_id_feat;
+
+        uint32_t num_mistakes;
+        uint32_t num_ecs;
+        uint32_t num_test_ecs;
+        uint32_t test_mistakes;
+        int learn_at_leaf;
+
+        bool test_mode;
+
+        size_t current_pass;
+        size_t final_pass;
+
+        int top_K;
+        int oas;
+	    int dream_at_update;
+
+        int online;
+
+        float F1_score;
+        float hamming_loss;
+
+        memory_tree()
+        {
+            nodes = v_init<node>();
+            examples = v_init<example*>();
+            alpha = 0.5;
+            routers_used = 0;
+            iter = 0;
+            num_mistakes = 0;
+            num_ecs = 0;
+            num_test_ecs = 0;
+            path_id_feat = false;
+            test_mode = false;
+            max_depth = 0;
+            max_ex_in_leaf = 0;
+            construct_time = 0;
+            test_time = 0;
+            top_K = 1;
+        }
+    };
+
+    float linear_kernel(const flat_example* fec1, const flat_example* fec2)
+    { 
+        float dotprod = 0;
+   
+        features& fs_1 = (features&)fec1->fs;
+        features& fs_2 = (features&)fec2->fs;
+        if (fs_2.indicies.size() == 0)
+            return 0.f;
+   
+        int numint = 0;
+        for (size_t idx1 = 0, idx2 = 0; idx1 < fs_1.size() && idx2 < fs_2.size() ; idx1++)
+        { 
+            uint64_t ec1pos = fs_1.indicies[idx1];
+            uint64_t ec2pos = fs_2.indicies[idx2];
+            //params.all->trace_message<<ec1pos<<" "<<ec2pos<<" "<<idx1<<" "<<idx2<<" "<<f->x<<" "<<ec2f->x<<endl;
+            if(ec1pos < ec2pos) continue;
+   
+            while(ec1pos > ec2pos && ++idx2 < fs_2.size())
+                ec2pos = fs_2.indicies[idx2];
+   
+            if(ec1pos == ec2pos)
+            { //params.all->trace_message<<ec1pos<<" "<<ec2pos<<" "<<idx1<<" "<<idx2<<" "<<f->x<<" "<<ec2f->x<<endl;
+                numint++;
+                dotprod += fs_1.values[idx1] * fs_2.values[idx2];
+                ++idx2;
+            }
+        }
+        return dotprod;
+   }
+
+
+    float normalized_linear_prod(memory_tree& b, example* ec1, example* ec2)
+    {
+        flat_example* fec1 = flatten_sort_example(*b.all, ec1);
+        flat_example* fec2 = flatten_sort_example(*b.all, ec2);
+        float linear_prod = linear_kernel(fec1, fec2);
+        fec1->fs.delete_v(); 
+        fec2->fs.delete_v();
+        free(fec1);
+        free(fec2);
```

Comment:
suggestion? I'm trying to compute <x1, x2> here. I created two flatten sorted examples from x1 & x2, and then do the linear product by looping over two sorted features.

Reply:
If we force sort_features, can all of the sorting code here go away?

Reply:
I don't think we want to force sort_features here because that's incompatible with ngrams (for example).

The way to handle this is via amortization.  In particular, we should be able to always store examples in the leaves in a flattened/sorted way and then flatten/sort any query once at the root, retaining the flattened form throughout the query.